ludwig/models/llm.py

```diff
@@ -480,13 +480,17 @@ def update_metrics(self, targets, predictions):
         """Updates the model's metrics given targets and predictions for zero-shot/few-shot."""
         for of_name, of_obj in self.output_features.items():
             if isinstance(of_obj, TextOutputFeature):
-                # Align the target length with the predictions length to enable text metric evaluation.
-                _targets, _predictions = realign_target_and_prediction_tensors_for_inference(
-                    targets, predictions, of_name, self.tokenizer
-                )
-                of_obj.update_metrics(_targets[of_name], _predictions[of_name])
-                continue
-            of_obj.update_metrics(targets[of_name], predictions[of_name])
+                # 1. Align the target length with the predictions length to enable text metric evaluation.
+                # 2. Decode both the target and predictions tokens to strings to calculate additional metrics.
+                (
+                    _targets,
+                    _decoded_targets,
+                    _predictions,
+                    _decoded_predictions,
+                ) = realign_target_and_prediction_tensors_for_inference(targets, predictions, of_name, self.tokenizer)
+                of_obj.update_metrics(_targets[of_name], _predictions[of_name], _decoded_targets, _decoded_predictions)
```

Comment:
do we streictly need to pass all this stuff? plus isn't there a more elegant way that having an if textfeature and having update_metrics to have a different signature only in that case? It's not amazing abstraction, and requires us to check if textoutput everywhere we call update metrics (it may be happening only here, but still, it's a red flag that comething is wrong with the abstraction). One option could be to have kwargs in the superclass update metrics and put the edditional things as parameters passed through kwargs. it would still need an if for determining the inputs, but at least we would not have textoutputs need a different signature for update_metrics

Reply:
Having `TextOutputFeature` have access to its tokenizer (for decoding predictions) would enable the original `update_metrics()` signature to work, though `TextOutputFeature` would still need to override the base class's implementation of `update_metrics()` to decode texts with the tokenizer and use the decoded texts to compute `RESPONSE` metrics.

However, constructing `TextOutputFeature` with the tokenizer would require an even more invasive change.
- Output feature construction code paths are still fully unified under `BaseModel.build_outputs()`. 
- We would need to change the `build_outputs()` signature to take in an additional `training_set_metadata`  argument (which contains tokenizer information), so that it can be passed through the output feature constructors, even though TextOutputFeature is the only output feature who would need training_set_metadata in this way.
- Output feature construction call sites would need to be updated as well -- there are 3 (`ecd.py`, `gbm.py`, and `llm.py`).

Reply:
This refactoring seems worthwhile to do if we anticipate `TextOutputFeature` being used for multiple model types. For now, keeping the ownership of the tokenizer at the model level, relying on the model to decode predictions, and provide them for `TextOutputFeature`'s metrics calculations seems reasonable to proceed with -- WDYT?

Reply:
I made one simplification, which is to do decoding within `update_metrics()` instead of outside at the model level. This should reduce some duplication.

Reply:
sounds good to me!